python/thunder/utils/serializable.py

```diff
@@ -0,0 +1,410 @@
+                    "shape": data.shape,
+                    "values": b64encode(data),
+                    "dtype":  str(data.dtype)}}
+        elif isinstance(data, Serializable):
+            # nested serializable object
+            return {"py/Serializable": {
+                "type": dataType.__name__,
+                "module": dataType.__module__,
+                "data": data.serialize()
+            }}
+
+        raise TypeError("Type %s not data-serializable" % dataType)
+
+    def serialize(self, numpyStorage='auto'):
+        """
+        Serialize this object to a python dictionary that can easily be converted
+        to/from JSON using Python's standard JSON library.
+
+        Parameters
+        ----------
+
+          numpyStorage: {'auto', 'ascii', 'base64' }, optional, default 'auto'
+            Use to select whether numpy arrays will be encoded in ASCII (as
+            a list of lists) in Base64 (i.e. space efficient binary), or to
+            select automatically (the default) depending on the size of the
+            array. Currently the Base64 encoding is selected if the array
+            has more than 1000 elements.
+
+        Returns
+        -------
+
+          The object encoded as a python dictionary with "JSON-safe" datatypes that is ready to
+          be converted to a string using Python's standard JSON library (or another library of
+          your choice).
+
+        """
+        # Check for unsupported class.
+        # a mix of slots and dicts can happen from multiple inheritance
+        # at the moment, this appears to be "working" - with the restriction that if there
+        # is both __slots__ and __dict__, only the __slots__ attributes will be serialized / deserialized.
+        # if hasattr(self, "__slots__") and hasattr(self, "__dict__"):
+        #    raise TypeError("Cannot serialize a class that has attributes in both __slots__ and __dict__")
+
+        # If this object has slots, we need to convert the slots to a dict before serializing them.
+        if hasattr(self, "__slots__"):
+            retVal = {}
+            for k in self.__slots__:
+                if hasattr(self, k):
+                    v = getattr(self, k)
+                    retVal[self.__serializeRecursively(k, numpyStorage)] = \
+                        self.__serializeRecursively(v, numpyStorage)
+            return retVal
+
+        # Otherwise, we handle the object as though it has a normal __dict__ containing its attributes.
+        else:
+            retVal = {}
+            for k, v in self.__dict__.iteritems():
+                retVal[self.__serializeRecursively(k, numpyStorage)] = \
+                    self.__serializeRecursively(v, numpyStorage)
+            return retVal
+
+    @classmethod
+    def deserialize(cls, serializedDict):
+        """
+        Restore the object that has been converted to a python dictionary using an @serializable
+        class's serialize() method.
+
+        Parameters
```

Comment:
Docstring formatting here is a bit inconsistent with the rest of the library in terms of spacing / line breaks.